src/apps/user_auth/pipeline.py

```diff
@@ -0,0 +1,52 @@
+from .models import User
+
+
+def get_username(strategy, uid, user=None, *args, **kwargs):
+    """If no user, set username to UID."""
+    if not user:
+        username = uid
+    else:
+        username = strategy.storage.user.get_username(user)
+    return {'username': username}
+
+
+def user_details(user, details, strategy, *args, **kwargs):
+    """Update user details using data from provider."""
+    if user:
+        changed = False  # flag to track changes
+        protected = ('id', 'pk') + tuple(strategy.setting('PROTECTED_USER_FIELDS', []))
+        user_attrs = [
+            'login' ,
+            'avatar_url',
+            'url',
+            'html_url',
+            'name',
+            'company',
+            'bio',
+            'email',
+        ]
+
+        if kwargs:
+            try:
+                response = kwargs.pop("response")
+                for attr in user_attrs:
+                    user_attr_value = getattr(user, attr, None)
+                    github_attr_value = response[attr]
+                    if not user_attr_value or attr not in protected:
+
+                        setattr(user, attr, github_attr_value)
+                setattr(user, 'uid', response['id'])
+                user.save()
+            except AttributeError:
+                print("Attribute not found.")
+        if changed:
+            strategy.storage.user.changed(user)
+
+
+def associate_existing_user(uid, *args, **kwargs):
+    """If there already is an user with the given uid, hand it over to the pipeline"""
```

Comment:
`an user` -> `a user` I think?

This is very cool

Reply:
I believe the rule of thumb for grammar in this case is if the noun begins with a vowel or not. Since U is a vowel it's `an` I believe.

Reply:
https://www.quora.com/Is-it-an-user-or-a-user

Think we can just do "a user" if the vowel sounds like a "y" ??! didn't know that!